test-data/unit/check-narrowing.test

```diff
@@ -445,3 +528,225 @@ if y["model"]["key"] is Key.C:
+
+[case testNarrowingEqualityFlipFlop]
+# flags: --warn-unreachable --strict-equality
+from typing_extensions import Literal, Final
+from enum import Enum
+
+class State(Enum):
+    A = 1
+    B = 2
+
+class FlipFlopEnum:
+    def __init__(self) -> None:
+        self.state = State.A
+
+    def mutate(self) -> None:
+        self.state = State.B if self.state == State.A else State.A
+
+class FlipFlopStr:
+    def __init__(self) -> None:
+        self.state = "state-1"
+
+    def mutate(self) -> None:
+        self.state = "state-2" if self.state == "state-1" else "state-1"
+
+def test1(switch: FlipFlopEnum) -> None:
+    # Naively, we might assume the 'assert' here would narrow the type to
+    # Literal[State.A]. However, doing this ends up breaking a fair number of real-world
+    # code (usually test cases) that looks similar to this function: e.g. checks
+    # to make sure a field was mutated to some particular value.
+    #
+    # And since mypy can't really reason about state mutation, we take a conservative
+    # approach and avoid narrowing anything here.
+
+    assert switch.state == State.A
+    reveal_type(switch.state)          # N: Revealed type is '__main__.State'
+
+    switch.mutate()
+
+    assert switch.state == State.B
+    reveal_type(switch.state)          # N: Revealed type is '__main__.State'
+
+def test2(switch: FlipFlopEnum) -> None:
+    # So strictly speaking, we ought to do the same thing with 'is' comparisons
+    # for the same reasons as above. But in practice, not too many people seem to
+    # know that doing 'some_enum is MyEnum.Value' is idiomatic. So in practice,
+    # this is probably good enough for now.
+
+    assert switch.state is State.A
+    reveal_type(switch.state)          # N: Revealed type is 'Literal[__main__.State.A]'
+
+    switch.mutate()
+
+    assert switch.state is State.B     # E: Non-overlapping identity check (left operand type: "Literal[State.A]", right operand type: "Literal[State.B]")
+    reveal_type(switch.state)          # E: Statement is unreachable
+
+def test3(switch: FlipFlopStr) -> None:
+    # This is the same thing as 'test1', except we try using str literals.
+
+    assert switch.state == "state-1"
+    reveal_type(switch.state)          # N: Revealed type is 'builtins.str'
+
+    switch.mutate()
+
+    assert switch.state == "state-2"
+    reveal_type(switch.state)          # N: Revealed type is 'builtins.str'
+[builtins fixtures/primitives.pyi]
+
+[case testNarrowingEqualityRequiresExplicitStrLiteral]
+# flags: --strict-optional
+from typing_extensions import Literal, Final
+
+A_final: Final = "A"
+A_literal: Literal["A"]
+
+# Neither the LHS nor the RHS are explicit literals, so regrettably nothing
+# is narrowed here -- see 'testNarrowingEqualityFlipFlop' for an example of
+# why more precise inference here is problematic.
+x_str: str
+if x_str == A_final:
```

Comment:
I don't think I understand why literal type is "better" than a final name. IMO final names should behave exactly as if they were replaced by their values.

Reply:
If it's any consolation, we also won't narrow anything away if we do `if x_str == "A"` -- I updated the test case with an example.

Basically, I found trying to narrow when we do `x_str == "A"` often ended up causing too many false positives, which is why I settled on the heuristic of requiring either the LHS or the RHS to be a `Literal[...]` of some sort before we attempt narrowing.